### ISSUE TITLE
panic when watching a path, moving an existing file to it, then appending to it

### ISSUE THREAD

**Issue by malban** [jmagnuson/linemux#36]
### Environment
Ubuntu 18.04
rustc 1.52.1
linemux = "0.2"

### To replicate:

**source:**
```
    let mut lines = MuxedLines::new()?;
    lines.add_file(logfile).await?;
```

```
$ mv <existing file> <logfile path>
$ echo "test" >> <logfile path>
```

**backtrace:**
```
thread 'main' panicked at 'missing reader position', /home/malban/.cargo/registry/src/github.com-1ecc6299db9ec823/linemux-0.2.1/src/reader.rs:468:34
stack backtrace:
   0:     0x562edf644e80 - std::backtrace_rs::backtrace::libunwind::trace::h63b7a90188ab5fb3
                               at /rustc/9bc8c42bb2f19e745a63f3445f1ac248fb015e53/library/std/src/../../backtrace/src/backtrace/libunwind.rs:90:5
   1:     0x562edf644e80 - std::backtrace_rs::backtrace::trace_unsynchronized::h80aefbf9b851eca7
                               at /rustc/9bc8c42bb2f19e745a63f3445f1ac248fb015e53/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
   2:     0x562edf644e80 - std::sys_common::backtrace::_print_fmt::hbef05ae4237a4d72
                               at /rustc/9bc8c42bb2f19e745a63f3445f1ac248fb015e53/library/std/src/sys_common/backtrace.rs:67:5
   3:     0x562edf644e80 - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::h28abce2fdb9884c2
                               at /rustc/9bc8c42bb2f19e745a63f3445f1ac248fb015e53/library/std/src/sys_common/backtrace.rs:46:22
   4:     0x562edf664c7f - core::fmt::write::h3b84512577ca38a8
                               at /rustc/9bc8c42bb2f19e745a63f3445f1ac248fb015e53/library/core/src/fmt/mod.rs:1092:17
   5:     0x562edf642382 - std::io::Write::write_fmt::h465f8feea02e2aa1
                               at /rustc/9bc8c42bb2f19e745a63f3445f1ac248fb015e53/library/std/src/io/mod.rs:1572:15
   6:     0x562edf647085 - std::sys_common::backtrace::_print::h525280ee0d29bdde
                               at /rustc/9bc8c42bb2f19e745a63f3445f1ac248fb015e53/library/std/src/sys_common/backtrace.rs:49:5
   7:     0x562edf647085 - std::sys_common::backtrace::print::h1f0f5b9f3ef8fb78
                               at /rustc/9bc8c42bb2f19e745a63f3445f1ac248fb015e53/library/std/src/sys_common/backtrace.rs:36:9
   8:     0x562edf647085 - std::panicking::default_hook::{{closure}}::ha5838f6faa4a5a8f
                               at /rustc/9bc8c42bb2f19e745a63f3445f1ac248fb015e53/library/std/src/panicking.rs:208:50
   9:     0x562edf646b33 - std::panicking::default_hook::hfb9fe98acb0dcb3b
                               at /rustc/9bc8c42bb2f19e745a63f3445f1ac248fb015e53/library/std/src/panicking.rs:225:9
  10:     0x562edf64768d - std::panicking::rust_panic_with_hook::hb89f5f19036e6af8
                               at /rustc/9bc8c42bb2f19e745a63f3445f1ac248fb015e53/library/std/src/panicking.rs:591:17
  11:     0x562edf647227 - std::panicking::begin_panic_handler::{{closure}}::h119e7951427f41da
                               at /rustc/9bc8c42bb2f19e745a63f3445f1ac248fb015e53/library/std/src/panicking.rs:497:13
  12:     0x562edf64533c - std::sys_common::backtrace::__rust_end_short_backtrace::hce386c44bf47a128
                               at /rustc/9bc8c42bb2f19e745a63f3445f1ac248fb015e53/library/std/src/sys_common/backtrace.rs:141:18
  13:     0x562edf647189 - rust_begin_unwind
                               at /rustc/9bc8c42bb2f19e745a63f3445f1ac248fb015e53/library/std/src/panicking.rs:493:5
  14:     0x562edf4c8611 - core::panicking::panic_fmt::h2242888e8769cd33
                               at /rustc/9bc8c42bb2f19e745a63f3445f1ac248fb015e53/library/core/src/panicking.rs:92:14
  15:     0x562edf4c8473 - core::option::expect_failed::h5ed4263027530e3b
                               at /rustc/9bc8c42bb2f19e745a63f3445f1ac248fb015e53/library/core/src/option.rs:1321:5
  16:     0x562edf4fd762 - linemux::reader::MuxedLines::poll_next_line::h98ca17c67f2f411a
  17:     0x562edf4ec50f - main::{{closure}}::h992bf30f35d897bf
  18:     0x562edf4dc806 - tokio::park::thread::CachedParkThread::block_on::hde7794aeb531ce7e
  19:     0x562edf4e6667 - tokio::runtime::thread_pool::ThreadPool::block_on::h9b674a10a75ab920
  20:     0x562edf4d0ca8 - tokio::runtime::Runtime::block_on::h6d9ec9f86d4f6b62
  21:     0x562edf4e77c6 - main::had22f1cac64ae809
  22:     0x562edf4e3c03 - std::sys_common::backtrace::__rust_begin_short_backtrace::h2c108c9bd1ff7d1b
  23:     0x562edf4e3c19 - std::rt::lang_start::{{closure}}::hbc4bf6e65f2b81c0
  24:     0x562edf647a8a - core::ops::function::impls::<impl core::ops::function::FnOnce<A> for &F>::call_once::h44574effd2120c86
                               at /rustc/9bc8c42bb2f19e745a63f3445f1ac248fb015e53/library/core/src/ops/function.rs:259:13
  25:     0x562edf647a8a - std::panicking::try::do_call::h10b0bd4879c8dfb0
                               at /rustc/9bc8c42bb2f19e745a63f3445f1ac248fb015e53/library/std/src/panicking.rs:379:40
  26:     0x562edf647a8a - std::panicking::try::h60c6780d33419e92
                               at /rustc/9bc8c42bb2f19e745a63f3445f1ac248fb015e53/library/std/src/panicking.rs:343:19
  27:     0x562edf647a8a - std::panic::catch_unwind::h111f33e08c52e2ce
                               at /rustc/9bc8c42bb2f19e745a63f3445f1ac248fb015e53/library/std/src/panic.rs:431:14
  28:     0x562edf647a8a - std::rt::lang_start_internal::h126f2e09345dbfda
                               at /rustc/9bc8c42bb2f19e745a63f3445f1ac248fb015e53/library/std/src/rt.rs:51:25
  29:     0x562edf4ed952 - main
  30:     0x7f981032cbf7 - __libc_start_main
  31:     0x562edf4c8cfa - _start
  32:                0x0 - <unknown>
```

**Comment by jmagnuson** [comment]
Thank you!

In https://github.com/jmagnuson/linemux/pull/24 `Rename` events were directly mapped to unwatches to prevent races during log rotations, but this is problematic because the above case is only represented as `Rename` events. For example `mv foo.txt bar.txt` in Linux:

```
[src/events.rs:67] &res = Ok(
    Event {
        kind: Modify(
            Name(
                From,
            ),
        ),
        paths: [
            "/home/jon/src/jmagnuson/linemux/foo.txt",
        ],
        attr:tracker: Some(
            7880741,
        ),
        attr:flag: None,
        attr:info: None,
        attr:source: None,
    },
)
[src/events.rs:67] &res = Ok(
    Event {
        kind: Modify(
            Name(
                To,
            ),
        ),
        paths: [
            "/home/jon/src/jmagnuson/linemux/bar.txt",
        ],
        attr:tracker: Some(
            7880741,
        ),
        attr:flag: None,
        attr:info: None,
        attr:source: None,
    },
)
[src/events.rs:67] &res = Ok(
    Event {
        kind: Modify(
            Name(
                Both,
            ),
        ),
        paths: [
            "/home/jon/src/jmagnuson/linemux/foo.txt",
            "/home/jon/src/jmagnuson/linemux/bar.txt",
        ],
        attr:tracker: Some(
            7880741,
        ),
        attr:flag: None,
        attr:info: None,
        attr:source: None,
    },
)
```

no `Create` events appear, which is mostly what linemux [currently relies on](https://github.com/jmagnuson/linemux/blob/f1668879d22a902fdc0270e5488f26cf4d453c7e/src/reader.rs#L392).

Further complicating things, `notify` only emits `RenameMode::From` events in OSX for both source and destination (instead of the expected `::To` for dest), so this needs special consideration as well.

**Comment by jmagnuson** [comment]
@malban I have a fix PR up in #37, would you mind trying to verify that it fixes the issue for you?

**Comment by malban** [comment]
Looks like it works.  Thanks!

**Comment by jmagnuson** [comment]
Published [`0.2.2`](https://crates.io/crates/linemux/0.2.2) containing the above fix.
